### ISSUE TITLE
Optional Transfer of implicit column rowid

### ISSUE THREAD

**Issue by FireController1847** [techouse/sqlite3-to-mysql#8]
Hello, it's me again. Sorry to be bothering you so much >.<

Now I'm no expert in python so I probably won't be able to make a PR for this, but in sqlite there appears to be an implicit column called "rowid". I was wondering if adding this as an optional row to transfer would be something you'd be willing to do?

Thanks again for this software this has saved my butt countless times, especially in my transfers as I migrate some data to MySQL :)

**Comment by techouse** [comment]
Well, that's a quirky specific. Will see what I can do, but I doubt it will be any time soon.

**Comment by FireController1847** [comment]
No problem, no hurry :)

**Comment by soulseekah** [comment]
@FireController1847 I'm a bit confused. Is this an INT row called `rowid`? Or do you mean that the type of the column is ROWID? If it's the latter, then my 0-affinity suggestion in #9 is actually something that would work in your case.

If the former, can you please show your table schema and what you want it to look like in MySQL?

**Comment by techouse** [comment]
> @FireController1847 I'm a bit confused. Is this an INT row called `rowid`? Or do you mean that the type of the column is ROWID? If it's the latter, then my 0-affinity suggestion in #9 is actually something that would work in your case.
> 
> If the former, can you please show your table schema and what you want it to look like in MySQL?

I think @FireController1847 refers to this https://www.sqlite.org/rowidtable.html  - so an internal SQLite thing which gets added when you create a table with `WITHOUT ROWID`.

**Comment by soulseekah** [comment]
Thanks. Makes sense. Can we do something like add a flag `--rowid=id` which creates a column called `id INT AUTOINCREMENT` or something? Would that work?

**Comment by techouse** [comment]
I think that `INT` won't be adequate because `rowid` is a 64bit integer, which in MySQL needs to be translated to a `BIGINT`.

**Comment by techouse** [comment]
So in essence, if I do this, you want me to transfer the `rowid` as a separate column named `rowid` and add a unique constraint to it as it is outlined [here](https://www.sqlite.org/rowidtable.html)?

I mean this is a rather messy leftover from legacy SQLite code ... not even sure at this point why we'd be needing this tbh.

**Comment by techouse** [comment]
> Thanks. Makes sense. Can we do something like add a flag `--rowid=id` which creates a column called `id INT AUTOINCREMENT` or something? Would that work?

I don't think that's wise. If you read the [spec](https://www.sqlite.org/rowidtable.html) `rowid` is not even supposed to be used by any external app let alone relying on it. If I transfer it I would name it `rowid` and then it's up to the user to rename the column so whatever they like.

**Comment by FireController1847** [comment]
> > Thanks. Makes sense. Can we do something like add a flag `--rowid=id` which creates a column called `id INT AUTOINCREMENT` or something? Would that work?
> 
> I don't think that's wise. If you read the [spec](https://www.sqlite.org/rowidtable.html) `rowid` is not even supposed to be used by any external app let alone relying on it. If I transfer it I would name it `rowid` and then it's up to the user to rename the column so whatever they like.

Why make the user manually rename it? I don't see as to why it couldn't be depended on. Similarly it's really only there to give us the option if, say, some idiots decided they did depend on it, and now I need to transfer their data to a MySQL. By using this it allows us to have the option, and by using a simple argument like that we can also choose the row name. You may not want to depend on it in SQLite, but that doesn't mean you can't in MySQL because it is essentially equivalent to a column `id` with `AUTO_INCREMENT` as a primary key

**Comment by techouse** [comment]
> > I don't think that's wise. If you read the [spec](https://www.sqlite.org/rowidtable.html) `rowid` is not even supposed to be used by any external app let alone relying on it. If I transfer it I would name it `rowid` and then it's up to the user to rename the column so whatever they like.
> 
> Why make the user manually rename it? ~I don't see as to why it couldn't be depended on.~ (I get it now, it's for backwards compatibility, however if they are still supporting it so should we.) It's really only there to give us the option if, say, some idiots decided they did depend on it, and now I need to transfer their data to MySQL. By using this it allows us to have the option, and by using a simple argument like that we can also choose the row name. You may not want to depend on it in SQLite, but that doesn't mean you can't in MySQL because it is essentially equivalent to a column `id` with `AUTO_INCREMENT` as a primary key.
> 
> This is basically just covering another commonly found edge case, however stupid, so that we can still use this amazing program to transfer even the dumbest of databases to use modern standards.

Heh, if some idiots depend on something called `rowid` then it makes little sense for the tool to rename it. It's much easier for you to simply open up a database manager and rename those fields manually. Doing too much will make the tools ultra-complicated. Don't forget I need to test all of this stuff.

**Comment by FireController1847** [comment]
I guess I just don't understand how it would make it more complicated. We can't implicitly transfer it, it should be an option as most of the time you won't want it transferred over, I presume. And if you're going to transfer it, you've still got to hard-code the column name. You just replace said hard-coded name with the argument.

But, I'm not the one coding it, and however you'd like to do it I will, in the end, support :)

**Comment by techouse** [comment]
You're forgetting 2 BIG things here.

1. Column names have to be unique. You chose a column to be named the same as another column then what? Drop the original?

2. Foreign keys. Those things rely on column names. 


I've already coded it, however, I've used the safest approach possible:
- transfer the SQLite `rowid` as a column named `rowid` (unless the SQLite table does not have it)
- add a `UNIQUE` constraint to that `rowid` column
- do not add `AUTO_INCREMENT` to that `rowid` column

If you want any of that fancy renaming you'll have to do it manually. Shouldn't take a lot of time in any database admin tool.
